pkg/api/customization/workload/workload.go

```diff
@@ -36,33 +38,108 @@ func (a ActionWrapper) ActionHandler(actionName string, action *types.Action, ap
 	namespace, name := splitID(deployment.ID)
 	switch actionName {
 	case "rollback":
-		actionInput, err := parse.ReadBody(apiContext.Request)
+		input, err := handler.ParseAndValidateActionBody(apiContext, apiContext.Schemas.Schema(&projectschema.Version,
+			projectclient.DeploymentRollbackInputType))
 		if err != nil {
-			return err
+			return httperror.NewAPIError(httperror.InvalidBodyContent,
+				fmt.Sprintf("Failed to parse action body: %v", err))
 		}
-		revision, _ := convert.ToNumber(actionInput["revision"])
+		rollbackInput := &projectclient.DeploymentRollbackInput{}
+		if err := mapstructure.Decode(input, rollbackInput); err != nil {
+			return httperror.NewAPIError(httperror.InvalidBodyContent,
+				fmt.Sprintf("Failed to parse body: %v", err))
+		}
+		revision := fetchRevisionFor(apiContext, rollbackInput, namespace, name, deployment.WorkloadAnnotations["deployment.kubernetes.io/revision"])
+		logrus.Debugf("rollbackInput %v", revision)
+		if revision == "" {
+			return httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("ReplicaSet %s doesn't exist for deployment %s", rollbackInput.ReplicaSetID, deployment.ID))
+		}
+		revisionNum, _ := convert.ToNumber(revision)
+
 		data := map[string]interface{}{}
 		data["kind"] = "DeploymentRollback"
 		data["apiVersion"] = "extensions/v1beta1"
 		data["name"] = name
-		data["rollbackTo"] = map[string]interface{}{"revision": revision}
+		data["rollbackTo"] = map[string]interface{}{"revision": revisionNum}
 		deploymentRollback, _ := json.Marshal(data)
-
-		err = a.Client.Post().Prefix("apis/extensions/v1beta1/").Namespace(namespace).Resource("deployments").Name(name).SubResource("rollback").Body(deploymentRollback).Do().Error()
+		clusterName := a.ClusterManager.ClusterName(apiContext)
+		if clusterName == "" {
```

Comment:
can we validate clusterName/clusterContext in the very beginning of the rollback logic?